File: src/java/action/RunTestAction.java
package action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiJavaFile;
import com.intellij.psi.PsiMethod;
import com.intellij.psi.PsiParameter;
import org.jetbrains.annotations.NotNull;
import util.BaseException;
import util.PluginFileHelper;
import util.RequestInvokeUtil;
import util.TipsUtils;
import vo.MethodVo;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * @author linxixin@cvte.com
 * @since 1.0
 */
public class RunTestAction extends AnAction {
    private Logger log = Logger.getInstance(RunTestAction.class);

    @Override
    public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
        log.info("触发方法执行事件, 请求开始");
        PluginFileHelper fileHelper = new PluginFileHelper(anActionEvent);
        Project project = anActionEvent.getProject();

        TipsUtils.catchBaseException(() -> {
            PsiMethod positionMethod = fileHelper.getPositionMethod();
            if (positionMethod.getParameterList()
                              .getParametersCount() > 0) {
                throw new BaseException("测试方法不能有参数");
            }
            ;
            String packageName = ((PsiJavaFile) positionMethod.getContainingFile()).getPackageName();
            List<String> paramNames = Arrays.stream(positionMethod.getParameterList()
                                                                  .getParameters())
                                            .map(PsiParameter::getName)
                                            .collect(Collectors.toList());

            String qualifiedName = positionMethod.getContainingClass()
                                                 .getQualifiedName();
            MethodVo methodVo = new MethodVo(
                    packageName,
                    qualifiedName,
                    positionMethod.getName(),
                    paramNames
            );
            RequestInvokeUtil.requestInvoke(methodVo, project);
        });
    }

}

File: src/java/action/RunTestAction.kt
package action

import base.SettingProperty
import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.debugger.DebuggerManagerEx
import com.intellij.debugger.impl.DebuggerSession
import com.intellij.debugger.settings.DebuggerSettings
import com.intellij.debugger.ui.HotSwapStatusListener
import com.intellij.debugger.ui.HotSwapUI
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiMethod
import com.intellij.psi.PsiParameter
import com.intellij.psi.impl.source.PsiJavaFileImpl
import component.InitStartupActivity
import util.PluginFileHelper
import util.RequestInvokeUtil
import util.catchBaseException
import vo.BaseException
import vo.MethodVo


/**
 *
 * @author linxixin@cvte.com
 */


class RunTestAction : AnAction() {

    private val log: Logger = Logger.getInstance(this.javaClass)

    override fun actionPerformed(e: AnActionEvent) {
        log.info("触发方法执行事件, 请求开始")
        val fileHelper = PluginFileHelper(e!!)
        catchBaseException {
            val positionMethod = fileHelper.getPositionMethod() ?: throw BaseException("请在方法内执行")

            if (positionMethod.parameterList.parametersCount > 0) {
                throw BaseException("测试方法不能有参数")
            }
            if (InitStartupActivity.FileChange.isFileChange()) {
                FileDocumentManager.getInstance().saveAllDocuments()
                InitStartupActivity.FileChange.isChange = false
                val debuggerManager = DebuggerManagerEx.getInstanceEx(e.project)
                val session = debuggerManager.context.debuggerSession

                if (session != null && session.isAttached) {
//                session.isModifiedClassesScanRequired
                    HotSwapUI.getInstance(e.project).reloadChangedClasses(session, DebuggerSettings.getInstance().COMPILE_BEFORE_HOTSWAP, object : HotSwapStatusListener {
                        override fun onSuccess(sessions: MutableList<DebuggerSession>?) {
                            ApplicationManager.getApplication().runReadAction {
                                invoke(positionMethod, e.project!!)
                            }
                        }
                    })

                }

            } else {
                invoke(positionMethod, e.project!!)
            }

        }
    }

    fun invoke(positionMethod: PsiMethod, project: Project) {

        val packageName = (positionMethod.containingFile as PsiJavaFileImpl).packageName
        val paramNames = positionMethod.parameterList.parameters.mapNotNull { psiParameter: PsiParameter? -> psiParameter!!.name }
        val methodVo = MethodVo(
                packageName,
                positionMethod.containingClass!!.qualifiedName!!,
                positionMethod.name,
                paramNames
        )
        RequestInvokeUtil.requestInvoke(methodVo, project)
        SettingProperty.setLastRequest(project, ObjectMapper().writeValueAsString(methodVo))
    }

}
